bindings/go: Reduce memory allocations when packing tuples

vmg opened the pull request:
Hi friendos! I was prototyping a FoundationDB layer using the Go bindings and I noticed a surprising amount of GC churn when having many concurrent transactions (100+). At first I thought it was related to the CGo interface, but it turns out it's the native Tuple packing code that was being a bit wasteful.

This PR improves the issue considerably, to the point where packing a key in the general case results in a single allocation.

- Commit dee5cf1 implements some more-or-less comprehensive test cases for Tuple packing. I'm aware the tuple packing functionality is tested through the `bindingtester` package, but I wanted a quicker way to iterate with some safety on the refactoring. Please let me know if you think these tests are redundant and I'll drop them! The tests also come with a benchmark for tuple packing so we can compare before and after.

- Commit 87c4ea4 performs some rather simple optimizations with surprisingly good results. There's nothing particularly complex here and nothing unsafe. The commit message includes a full description of all the implemented optimizations and their reasoning.

The end result is as follows:

```
benchmark                                  old ns/op     new ns/op     delta
BenchmarkTuplePacking/Simple-4             310           76.4          -75.35%
BenchmarkTuplePacking/Namespaces-4         495           137           -72.32%
BenchmarkTuplePacking/ManyStrings-4        960           255           -73.44%
BenchmarkTuplePacking/ManyStringsNil-4     1090          392           -64.04%
BenchmarkTuplePacking/ManyBytes-4          1409          399           -71.68%
BenchmarkTuplePacking/ManyBytesNil-4       1364          533           -60.92%
BenchmarkTuplePacking/LargeBytes-4         319           107           -66.46%
BenchmarkTuplePacking/LargeBytesNil-4      638           306           -52.04%
BenchmarkTuplePacking/Integers-4           2764          455           -83.54%
BenchmarkTuplePacking/Floats-4             3478          482           -86.14%
BenchmarkTuplePacking/Doubles-4            3654          575           -84.26%
BenchmarkTuplePacking/UUIDs-4              366           211           -42.35%

benchmark                                  old allocs     new allocs     delta
BenchmarkTuplePacking/Simple-4             6              1              -83.33%
BenchmarkTuplePacking/Namespaces-4         11             1              -90.91%
BenchmarkTuplePacking/ManyStrings-4        18             2              -88.89%
BenchmarkTuplePacking/ManyStringsNil-4     18             2              -88.89%
BenchmarkTuplePacking/ManyBytes-4          23             3              -86.96%
BenchmarkTuplePacking/ManyBytesNil-4       22             2              -90.91%
BenchmarkTuplePacking/LargeBytes-4         3              2              -33.33%
BenchmarkTuplePacking/LargeBytesNil-4      3              2              -33.33%
BenchmarkTuplePacking/Integers-4           63             3              -95.24%
BenchmarkTuplePacking/Floats-4             62             2              -96.77%
BenchmarkTuplePacking/Doubles-4            63             3              -95.24%
BenchmarkTuplePacking/UUIDs-4              2              2              +0.00%

benchmark                                  old bytes     new bytes     delta
BenchmarkTuplePacking/Simple-4             272           64            -76.47%
BenchmarkTuplePacking/Namespaces-4         208           64            -69.23%
BenchmarkTuplePacking/ManyStrings-4        512           192           -62.50%
BenchmarkTuplePacking/ManyStringsNil-4     512           192           -62.50%
BenchmarkTuplePacking/ManyBytes-4          864           448           -48.15%
BenchmarkTuplePacking/ManyBytesNil-4       336           192           -42.86%
BenchmarkTuplePacking/LargeBytes-4         400           192           -52.00%
BenchmarkTuplePacking/LargeBytesNil-4      400           192           -52.00%
BenchmarkTuplePacking/Integers-4           3104          448           -85.57%
BenchmarkTuplePacking/Floats-4             2656          192           -92.77%
BenchmarkTuplePacking/Doubles-4            3104          448           -85.57%
BenchmarkTuplePacking/UUIDs-4              256           192           -25.00%
```

I'm quite happy with the results: the numbers are obviously good, particularly in comparison to the complexity (or lack thereof) of the optimizations, and with this patch applied, key packing basically becomes noise when memory profiling my whole service, with 1 allocation on average per key.

Please let me know if there's something I can tweak to make landing this PR easier. Thank you!

alecgrieser commented on bindings/go/src/fdb/tuple/tuple.go:
```diff
@@ -106,109 +115,128 @@ func adjustFloatBytes(b []byte, encode bool) {
-	buf.Write(bytes.Replace(b, []byte{0x00}, []byte{0x00, 0xFF}, -1))
-	buf.WriteByte(0x00)
+type packer struct {
+	buf []byte
```
What's benefit of wrapping the byte slice in a struct (rather than just keeping a raw slice and calling append on it as necessary)? Is it just to have a handle wrapping the slice to make it mutable (or mutable in length)?

There was a question on the forums about what it would take to make golang support versionstamps (https://forums.foundationdb.org/t/golang-versionstamp-usage/253). I think the answer might be "add a field to the packer struct saying at what position you wrote a versionstamp that will be completed at commit time", so we might want it anyway.

vmg replied:
I went with a struct because it makes the code more ergonomic (i.e. we can afford to have a `p.putBytes` method because it always gets inlined). Looking at the version stamps issue, it seems like it's a good choice rather than passing around all the data through the call stack.

alecgrieser commented on bindings/go/src/fdb/tuple/tuple.go:
```diff
@@ -106,109 +115,128 @@ func adjustFloatBytes(b []byte, encode bool) {
-		binary.Write(&ibuf, binary.BigEndian, int64(sizeLimits[n])+i)
+		p.putbyte(byte(0x14 - n))
+		ii := int64(sizeLimits[n]) + i
+		binary.BigEndian.PutUint64(scratch[:], uint64(ii))
```
Hm, this is nit-picking (sorry), but is there a better name that `ii` for that variable? If it weren't obnoxious, it might make sense to call it, like `offsetEncodedI`, but that's somewhat of a mouthful.

vmg replied:
Fixed!

alecgrieser commented on bindings/go/src/fdb/tuple/tuple_test.go:
```diff
@@ -0,0 +1,119 @@
+var update = flag.Bool("update", false, "update .golden files")
+
+func loadGolden(t *testing.T) (golden map[string][]byte) {
+	f, err := os.Open("testdata/tuples.golden")
```
Does this break if the test is run from a directory other than the directory where the test file is located in the source tree?

vmg replied:
Tests in Go are always run on the directory they reside, so assets can be loaded trivially. This is lifted straight from the "golden files" testing pattern on the standard library:

https://golang.org/src/cmd/gofmt/gofmt_test.go#L134

vmg replied:
By "always run" I mean that `go test` enforces this, regardless of where you're running `go test` from. 😄

alecgrieser replied:
Oh, interesting, I didn't know that.

alecgrieser commented on bindings/go/src/fdb/tuple/tuple_test.go:
```diff
@@ -0,0 +1,119 @@
+	{"Doubles", mktuple(genDouble, 20)},
+	{"UUIDs", Tuple{testUUID, true, testUUID, false, testUUID, true, testUUID, false, testUUID, true}},
+	{"NilCases", Tuple{"\x00", "\x00\xFF", "\x00\x00\x00", "\xFF\x00", ""}},
+}
```
Is there a test case that tests nested tuples?

vmg replied:
I'll write one!

vmg commented on bindings/go/src/fdb/tuple/tuple.go:
```diff
@@ -106,109 +115,128 @@ func adjustFloatBytes(b []byte, encode bool) {
+	p.buf = append(p.buf, b...)
+}
+
+func (p *packer) putbytesNil(b []byte, i int) {
```
Yes. I dunno what I was thinking 😅

vmg commented:
Fixed up @alecgrieser!

- Rebased on top of `release-5.2`
- Capitalized `putBytes` et al, renamed `ii` to something more obvious
- Added a test case for nested tuple packing

Thanks for the review. Let me know what else would you like me to tweak!

vmg commented:
🙏🙇‍♂️
